pkg/apis/crd/v1alpha1/types.go

```diff
@@ -291,3 +291,151 @@ type TLSProtocol struct {
 	// SNI (Server Name Indication) indicates the server domain name in the TLS/SSL hello message.
 	SNI string `json:"sni,omitempty"`
 }
+
+// Source describes the source spec of the traceflow.
+type Source struct {
+	// Namespace is the source namespace.
+	Namespace string `json:"namespace,omitempty"`
+	// Pod is the source pod.
+	Pod string `json:"pod,omitempty"`
+	// IP is the source IPv4 or IPv6 address. IP as the source is supported
+	// only for live-traffic Traceflow.
+	IP string `json:"ip,omitempty"`
+}
+
+// Destination describes the destination spec of the traceflow.
+type Destination struct {
+	// Namespace is the destination namespace.
+	Namespace string `json:"namespace,omitempty"`
+	// Pod is the destination pod, exclusive with destination service.
+	Pod string `json:"pod,omitempty"`
+	// Service is the destination service, exclusive with destination pod.
+	Service string `json:"service,omitempty"`
+	// IP is the destination IPv4 or IPv6 address.
+	IP string `json:"ip,omitempty"`
+}
+
+// IPHeader describes spec of an IPv4 header.
+type IPHeader struct {
+	// Protocol is the IP protocol.
+	Protocol int32 `json:"protocol,omitempty" yaml:"protocol,omitempty"`
+	// TTL is the IP TTL.
+	TTL int32 `json:"ttl,omitempty" yaml:"ttl,omitempty"`
+	// Flags is the flags for IP.
+	Flags int32 `json:"flags,omitempty" yaml:"flags,omitempty"`
+}
+
+// IPv6Header describes spec of an IPv6 header.
+type IPv6Header struct {
+	// NextHeader is the IPv6 protocol.
+	NextHeader *int32 `json:"nextHeader,omitempty" yaml:"nextHeader,omitempty"`
+	// HopLimit is the IPv6 Hop Limit.
+	HopLimit int32 `json:"hopLimit,omitempty" yaml:"hopLimit,omitempty"`
+}
+
+// TransportHeader describes spec of a TransportHeader.
+type TransportHeader struct {
+	ICMP *ICMPEchoRequestHeader `json:"icmp,omitempty" yaml:"icmp,omitempty"`
+	UDP  *UDPHeader             `json:"udp,omitempty" yaml:"udp,omitempty"`
+	TCP  *TCPHeader             `json:"tcp,omitempty" yaml:"tcp,omitempty"`
+}
+
+// ICMPEchoRequestHeader describes spec of an ICMP echo request header.
+type ICMPEchoRequestHeader struct {
+	// ID is the ICMPEchoRequestHeader ID.
+	ID int32 `json:"id,omitempty"`
+	// Sequence is the ICMPEchoRequestHeader sequence.
+	Sequence int32 `json:"sequence,omitempty"`
+}
+
+// UDPHeader describes spec of a UDP header.
+type UDPHeader struct {
+	// SrcPort is the source port.
+	SrcPort int32 `json:"srcPort,omitempty"`
+	// DstPort is the destination port.
+	DstPort int32 `json:"dstPort,omitempty"`
```

Comment:
I don't know how @jianjuns and @tnqn will feel about this, but since this is a new API, I think we should follow the [K8s convention](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md#optional-vs-required):

> Therefore, we ask that pointers always be used with optional fields that do not have a built-in nil value.

It tends to lead to a lot of pointers, but it makes it clearer what is optional and what is not. When creating a resource with YAML, it makes no difference. When creating a resource in Go, we can use `ptr.To` from the K8s libraries to make initialization less verbose. What do you think?

Reply:
Agreed. There maybe a small problem if we upgrade to v1beta1 later to share the same structure with Traceflow though.

Reply:
I'd rather do the best interface possible now for `PacketCapture`, and worry about unifying later.

If we have to introduce a new `v1beta2` version for Traceflow, it should still be possible to define a conversion webhook to convert between `v1beta1` and `v1beta2`, because of the semantics of the field?

Reply:
updated. Can you help review this change? I'm not very sure if i have done it right.

Reply:
I think @antoninbas means:
```
type UDPHeader struct {
	// SrcPort is the source port.
	SrcPort *int32 `json:"srcPort,omitempty"`
	// DstPort is the destination port.
	DstPort *int32 `json:"dstPort,omitempty"`
}
```
because they are optional.

Reply:
Just saw newer comments. Please ignore it.